src/coverlet.core/Instrumentation/ModuleTrackerTemplate.cs

```diff
@@ -15,130 +16,133 @@ namespace Coverlet.Core.Instrumentation
+            // to initialize the static setup fields according to the values derived from the instrumentation of
             // the module.
+
+            if (DefaultHitsFilePath != null)
+                Singleton = new ModuleTrackerTemplate(DefaultHitsFilePath, DefaultHitsArraySize);
+
+            //we always keep the view accessor around without disposing it. So, regardless of what happens, be it the module unloading or the process terminating, it will close properly when the finalizers are run and, failing that, when the kernel does process cleanup
         }
 
         public static void RecordHit(int hitLocationIndex)
         {
-            if (t_isTracking)
-                return;
-
-            if (t_threadHits == null)
-            {
-                t_isTracking = true;
-                lock (_threads)
-                {
-                    if (t_threadHits == null)
-                    {
-                        t_threadHits = new int[HitsArray.Length];
-                        _threads.Add(t_threadHits);
-                    }
-                }
-                t_isTracking = false;
-            }
-
-            ++t_threadHits[hitLocationIndex];
+            if (Singleton == null)
+                throw new InvalidOperationException("Singleton not initialized!");
+            Singleton.InstanceRecordHit(hitLocationIndex);
         }
 
-        public static void UnloadModule(object sender, EventArgs e)
+        public ModuleTrackerTemplate(string hitsFilePath, int hitsArraySize)
         {
-            t_isTracking = true;
+            if (hitsFilePath == null)
+                throw new ArgumentNullException(nameof(hitsFilePath));
 
-            // Update the global hits array from data from all the threads
-            lock (_threads)
-            {
-                foreach (var threadHits in _threads)
-                {
-                    for (int i = 0; i < HitsArray.Length; ++i)
-                        HitsArray[i] += threadHits[i];
-                }
+            if (hitsArraySize < 0)
+                throw new ArgumentOutOfRangeException(nameof(hitsArraySize), hitsArraySize, "hitsArraySize must not be less than 0!");
 
-                // Prevent any double counting scenario, i.e.: UnloadModule called twice (not sure if this can happen in practice ...)
-                // Only an issue if DomainUnload and ProcessExit can both happens, perhaps can be removed...
-                _threads.Clear();
-            }
+            this.hitsArraySize = hitsArraySize;
 
-            // The same module can be unloaded multiple times in the same process via different app domains.
-            // Use a global mutex to ensure no concurrent access.
-            using (var mutex = new Mutex(true, Path.GetFileNameWithoutExtension(HitsFilePath) + "_Mutex", out bool createdNew))
+            //now HitsFilePath and HitsArraySize should be populated
+
+            //first, try to create the hits file if it doesn't exist already
```

Comment:
Could you use https://docs.microsoft.com/en-us/dotnet/api/system.io.memorymappedfiles.memorymappedfile.createnew?view=netcore-2.1#System_IO_MemoryMappedFiles_MemoryMappedFile_CreateNew_System_String_System_Int64_ instead? The file doesn't actually have to contain all zeros, the OS will create zeroed memory pages as necessary.